### PR TITLE
Allow to set a different number of MAC retransmissions for different classes of packets

### PR DESCRIPTION
**Pull request by atiselsts**
Solves #106.

This patch adds support for setting transmission limits for UDP and TCP connnections. I'm open to also doing if for ICMP and/or RPL packets, but then we need to discuss the API (some sort of `#define` for RPL, which is then used as a parameter for ICMP functions called by RPL?)

**Edit:** needs conditional compilation with `UIP_CONF_WITH_VARIABLE_RETRANSMISSIONS` set to nonzero.

Usage example for a UDP application:

```
  struct simple_udp_connection simple_udp_conn;
  simple_udp_register(&simple_udp_conn, UDP_CLIENT_PORT, NULL,
                      UDP_SERVER_PORT, udp_rx_callback);
  /* Set the MAC max transmission limit to 13 */
  uip_set_max_mac_transmissions(simple_udp_conn.udp_conn, 13);
```

Usage example for a TCP application:

```
  struct tcp_socket socket;
  tcp_socket_connect(&socket, &address, port);
  if(socket.c != NULL) {
    /* Set the MAC max transmission limit to 13 */
    uip_set_max_mac_transmissions(socket.c, 13);
  }
```

### CODE REVIEW

File: os/net/mac/csma/csma-output.c [hunk 1]
```diff
@@ -85,9 +85,9 @@
 
 /* macMaxFrameRetries: Maximum number of re-transmissions attampts. Range 0--7 */
 #ifdef CSMA_CONF_MAX_FRAME_RETRIES
-#define CSMA_MAX_MAX_FRAME_RETRIES CSMA_CONF_MAX_FRAME_RETRIES
+#define CSMA_MAC_MAX_FRAME_RETRIES CSMA_CONF_MAX_FRAME_RETRIES
 #else
-#define CSMA_MAX_MAX_FRAME_RETRIES 7
+#define CSMA_MAC_MAX_FRAME_RETRIES 7
```
Comment (simonduq) [inline os/net/mac/csma/csma-output.c]: maybe `CSMA_MAX_FRAME_RETRIES`?

File: os/net/ipv6/uipopt.h [hunk 2]
```diff
@@ -518,6 +518,35 @@ void uip_log(char *msg);
...
 #define UIP_DEFAULT_PREFIX_LEN 64
 
+/**
+ * Enables selection of maximal MAC-layer transmission count at application layer
+ */
+#ifdef UIP_CONF_WITH_VARIABLE_RETRANSMISSIONS
+#define UIP_WITH_VARIABLE_RETRANSMISSIONS UIP_CONF_WITH_VARIABLE_RETRANSMISSIONS
+#else
+#define UIP_WITH_VARIABLE_RETRANSMISSIONS 0
+#endif
```
Comment (simonduq) [inline os/net/ipv6/uipopt.h]: do we really need to make this optional?

Comment (atiselsts) [reply os/net/ipv6/uipopt.h]: `.rom` overflow on Sky otherwise.

File: os/net/mac/csma/csma-output.c (hunk 1 above)
Comment (atiselsts) [reply os/net/mac/csma/csma-output.c]: I changed this to the current name to be consistent with TSCH. Tried grepping `MAX_MAC` in CSMA code, found nothing. With your suggest name I would still find nothing...

**Comment by atiselsts** [comment]
Thanks for the review Simon. Please suggest API for ICMP if you want to this feature. One way would be to add an argument to `uip_icmp6_send`.

Actually using that API from RPL better would fit in a separate PR.

**Comment by simonduq** [comment]
Yes, also works for me in a separate PR. New arg to `uip_icmp6_send` sounds good

File: os/net/mac/csma/csma-output.c (hunk 1 above)
Comment (simonduq) [reply os/net/mac/csma/csma-output.c]: OK I don't really mind about the name but let's keep it consistent with the `_CONF_` version of it

File: os/net/ipv6/uipopt.h (hunk 2 above)
Comment (simonduq) [reply os/net/ipv6/uipopt.h]: argh, that is going to be annoying soon enough :s

File: os/net/mac/csma/csma-output.c (hunk 1 above)
Comment (simonduq) [reply os/net/mac/csma/csma-output.c]: Hmm I didn't notice there was already an inconsistency :p would be good to fix anyway, right?

**Pull request merged**